presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -89,400 +111,733 @@
+            baseTable = (Table) aliasedRelation.getRelation();
+            return rewriteQuerySpecificationIfCompatible(node, baseTable);
+        }
 
-            ImmutableList.Builder<Field> fields = ImmutableList.builder();
+        if (!node.getFrom().isPresent()) {
+            return node;
+        }
 
-            for (ColumnHandle columnHandle : metadata.getColumnHandles(session, tableHandle.get()).values()) {
-                ColumnMetadata columnMetadata = metadata.getColumnMetadata(session, tableHandle.get(), columnHandle);
-                fields.add(Field.newUnqualified(materializedViewInfo.getWhereClause().get().getLocation(), columnMetadata.getName(), columnMetadata.getType()));
-            }
+        Relation newFrom = processSameType(from);
+        if (from == newFrom) {
+            return node;
+        }
 
-            Scope scope = Scope.builder()
-                    .withRelationType(RelationId.anonymous(), new RelationType(fields.build()))
-                    .build();
-
-            // Given base query's filter condition and materialized view's filter condition, the goal is to check if materialized view's
-            // filters contain Base's filters (Base implies materialized view).
-            // Let base query's filter condition be A, and materialized view's filter condition be B.
-            // One way to evaluate A implies B is to evaluate logical expression A^~B and check if the output domain is none.
-            // If the resulting domain is none, then A^~B is false. Thus A implies B.
-            // For more information and examples: https://fb.quip.com/WwmxA40jLMxR
-            // TODO: Implement method that utilizes external SAT solver libraries. https://github.com/prestodb/presto/issues/16536
-            RowExpression materializedViewWhereCondition = convertToRowExpression(materializedViewInfo.getWhereClause().get(), scope);
-            RowExpression baseQueryWhereCondition = convertToRowExpression(node.getWhere().get(), scope);
-            RowExpression rewriteLogicExpression = and(baseQueryWhereCondition,
-                    call(baseQueryWhereCondition.getSourceLocation(),
-                            "not",
-                            new FunctionResolution(metadata.getFunctionAndTypeManager()).notFunction(),
-                            materializedViewWhereCondition.getType(),
-                            materializedViewWhereCondition));
-            RowExpression disjunctiveNormalForm = logicalRowExpressions.convertToDisjunctiveNormalForm(rewriteLogicExpression);
-            ExtractionResult<VariableReferenceExpression> result = domainTranslator.fromPredicate(session.toConnectorSession(), disjunctiveNormalForm, BASIC_COLUMN_EXTRACTOR);
-
-            if (!result.getTupleDomain().equals(TupleDomain.none())) {
-                throw new IllegalStateException("View filter condition does not contain base query's filter condition");
-            }
+        if (node.getLocation().isPresent()) {
```

Comment:
We can use just one constructor, by passing location empty if not present?

Reply:
I'm going to resolve this and we can unify discussion https://github.com/prestodb/presto/pull/17901#discussion_r909102690